sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/datagen/DataGeneratorTableProviderTest.java

```diff
@@ -0,0 +1,296 @@
+
+      Assert.assertTrue("past_timestamp should be in the past", pastTimestamp.isBefore(now));
+      Assert.assertTrue(
+          "now_timestamp should be very recent",
+          now.plus(Duration.millis(100)).isAfter(nowTimestamp));
+    }
+  }
+
+  @Test
+  public void testAllDataTypes() {
+    String createDdl =
+        "CREATE EXTERNAL TABLE all_types_table (\n"
+            + "  is_active BOOLEAN,\n"
+            + "  cost DECIMAL,\n"
+            + "  past_timestamp TIMESTAMP,\n"
+            + "  now_timestamp TIMESTAMP\n"
+            + ") TYPE 'datagen' TBLPROPERTIES '{\n"
+            + "  \"number-of-rows\": \"10\",\n"
+            + "  \"fields.cost.min\": \"10.50\",\n"
+            + "  \"fields.cost.max\": \"99.99\",\n"
+            + "  \"fields.past_timestamp.max-past\": \"3600000\"\n"
+            + "}'";
+
+    pipeline
+        .apply(
+            "testAllDataTypes",
+            SqlTransform.query("SELECT * FROM all_types_table").withDdlString(createDdl))
+        .apply("ValidateAllTypes", ParDo.of(new ValidateAllTypesFn()));
+
+    pipeline.run().waitUntilFinish();
+  }
+
+  @Test
+  public void testMissingRequiredPropertyThrowsException() {
+    String createDdl = "CREATE EXTERNAL TABLE bad_table (id INT) TYPE 'datagen' TBLPROPERTIES '{}'";
+
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage(
+        "A 'datagen' table requires either 'rows-per-second' (for unbounded) or 'number-of-rows' (for bounded) in TBLPROPERTIES.");
+
+    pipeline.apply(
+        "testMissingRequiredProperty",
+        SqlTransform.query("SELECT * FROM bad_table").withDdlString(createDdl));
+    pipeline.run().waitUntilFinish();
+  }
+
+  /**
+   * Tests the default processing-time behavior to ensure it still works correctly and that
+   * event-time configuration is not required.
+   */
+  @Test
+  public void testProcessingTimeBehavior() {
+    String createDdl =
+        "CREATE EXTERNAL TABLE processing_time_table (\n"
+            + "  id BIGINT\n"
+            + ") TYPE 'datagen' TBLPROPERTIES '{\n"
+            + "  \"number-of-rows\": \"20\"\n"
+            + "}'";
+
+    PCollection<Row> result =
+        pipeline.apply(
+            "testProcessingTimeBehavior",
+            SqlTransform.query("SELECT * FROM processing_time_table").withDdlString(createDdl));
+
+    PAssert.that(result.apply("CountRows", Count.globally())).containsInAnyOrder(20L);
+
+    pipeline.run().waitUntilFinish();
+  }
+
+  private static class ValidateMultiTimestampFn extends DoFn<Row, Void> {
+    @ProcessElement
+    public void processElement(@Element Row row) {
+      Instant mainEventTime = row.getDateTime("main_event_time").toInstant();
+      Instant secondaryTime = row.getDateTime("secondary_time").toInstant();
+
+      Assert.assertTrue(
+          "Secondary timestamp should be recent",
+          Instant.now().plus(Duration.millis(200)).isAfter(secondaryTime));
```

Comment:
I think we can also check that the element timestamp is indeed equal to `main_event_time`
```suggestion
    public void processElement(@Element Row row, @Timestamp Instant rowTs) {
      Instant mainEventTime = row.getDateTime("main_event_time").toInstant();
      Instant secondaryTime = row.getDateTime("secondary_time").toInstant();

      Assert.assertEquals(mainEventTime, rowTs);
```